### DISCUSSION TITLE
Can someone help me explain the GPU allocation algorithm under the besteffort strategy?

### DISCUSSION THREAD

**Discussion by gcx-17211270** [NVIDIA/k8s-device-plugin#820] [Q&A]
https://github.com/NVIDIA/go-gpuallocator/blob/12c3885850231c64f7f67855ff581ebd6e7465bd/gpuallocator/besteffort_policy.go#L33

The Allocate method's comment describes: "it looks for a solution that maximizes the total score when dividing up all available GPUs on the node into sets of size 'size' and then summing their individual scores." 

It further explains: "Such a solution is necessary in the general case because of the non-hierarchical nature of the various links that influence the score calculated for each pair of GPUs." 

However, I still don't understand why this is necessary. Can someone provide a detailed explanation or give an example to illustrate why failing to do so would result in a suboptimal outcome?

---
有人可以帮我解释下besteffort策略下的GPU卡分配算法吗？在Allocate方法的注释上描述了：it looks for a solution that maximizes the total score when dividing up all available GPUs on the node into sets of size 'size' and then summing their individual scores.并在其后说明：Such a solution is necessary in the general case because of the non-hierarchical nature of the various links that influence the score calculated for each pair of GPUs. 但是我还是不明白为什么这样是必要的？有人能给我详细说明，或者举个例子说明为什么不这样做就无法得到最优的结果吗？

**Comment by ajavanma** [comment]
The goal is to avoid choosing a very good group now that leaves a very poor group for the next allocation of the same size.

For example, take four GPUs `A, B, C, D`, `size=2`, and no required GPUs. Suppose their pair scores are:

```text
AB = 300    CD = 10
AC = 200    BD = 200
AD = 10     BC = 10
```

These are illustrative scores, not measurements of a particular machine. There are only three ways to divide those GPUs into disjoint pairs:

| Partition | Total score |
| --- | ---: |
| `{A,B}`, `{C,D}` | 310 |
| `{A,C}`, `{B,D}` | 400 |
| `{A,D}`, `{B,C}` | 20 |

A greedy choice returns `{A,B}` because 300 is the highest individual score, leaving `{C,D}` with only 10. The partition-based choice selects the second row, then returns one of its 200-point pairs. The other 200-point pair remains available.

That is the significance of “non-hierarchical”: the PCIe hierarchy is not the only input. NVLink connections also contribute to the pair scores, so the best individual pair need not belong to the best overall partition.

In the revision you linked, [`Allocate`](https://github.com/NVIDIA/go-gpuallocator/blob/12c3885850231c64f7f67855ff581ebd6e7465bd/gpuallocator/besteffort_policy.go#L53-L96) first chooses the highest-scoring partition, then selects a set within it that contains all required GPUs. [`calculateGPUSetScore` and `calculateGPUPartitionScore`](https://github.com/NVIDIA/go-gpuallocator/blob/12c3885850231c64f7f67855ff581ebd6e7465bd/gpuallocator/besteffort_policy.go#L376-L398) implement those two sums.

This optimizes the allocator's topology score for equal-sized groups. It does not guarantee maximum application throughput or an optimal allocation for arbitrary future request sizes.
